src/reanimated2/animation/clamp.ts

```diff
@@ -0,0 +1,92 @@
+'use strict';
+import { defineAnimation, getReduceMotionForAnimation } from './util';
+import type {
+  Animation,
+  Timestamp,
+  AnimatableValue,
+  AnimationObject,
+  ReduceMotion,
+} from '../commonTypes';
+import type { ClampAnimation } from './commonTypes';
+
+type withClampType = <T extends AnimatableValue>(
+  clamp: { min: number; max: number },
+  delayedAnimation: T,
+  reduceMotion?: ReduceMotion
+) => T;
```

Comment:
1. I don't have a strong opinion here but I think that the animation should be the first argument. Indeed, in `withDelay`, the animation is passed as second argument and the first argument is the delay, however in `withRepeat` the number of repetitions is declared after the animation. In this case, the clamp is applied "after" the animation yields the value, so this would correspond better with the order of params.
2. Both `min` and `max` should be optional parameters so that the animation can be clamped from one side only.
3. I don't like the fact that we pass `reduceMotion` as yet another argument. Maybe we can group `min`, `max` and `reduceMotion` in a single `config` object, like we do for spring and timing?
4. I'm wondering how `withClamp` would work for colors 😄 Can we come up with some DX (developer experience) tricks to ensure that it's used only for numbers (e.g. enforce it using TypeScript)?

Reply:
1. You are right that it is justified to have the animation as the first parameter. On the other hand people need `withDelay` much more often than `withRepeat` and I guess that they are more familiar with syntax with animation being the second, so I think I would keep it like that.
2. Of course!
3. Sure! (BTW. Wouldn't it be better not to pass `reduceMotion` anywhere, but to have  higher order animation withReduceMotion?)
4. Colours have some weird bugs with spring. Generally animation between two colours works worse than animation single numerical progress and interpolating that color.
Yes, we can make sure that user provides number

Reply:
1. Let's discuss it with the whole team
2. Thank you!
3. Thank you! I think we were considering it at some point but please ask @bartlomiejbloniarz for details.
4. Thank you Some time ago we greatly improved the implementation of `interpolateColor`. I thought `withSpring(color)` also works fine, are you able to prepare a repro of the problem you've mentioned?

Reply:
ad 3. we considered it at some point, but decided that adding a reduced motion prop / config would be a smaller change to the api convention